### PR TITLE
[`Add`] Transaction Builder

### PR DESCRIPTION
**Pull request by cschuchardt88**
# Description
Easy simple builders for creating with signing, along with creating scripts for transactions.

## Change Log
- Add new `Transaction`, `Witness`, `Signer`, `TransactionAttribute`, `WitnessRule` and `WitnessCondition` builders

## Example
```csharp
var myPublicKey = ECPoint.Parse("021821807f923a3da004fb73871509d7635bcc05f41edef2a3ca5c941d8bbc1231", ECCurve.Secp256r1);
var tx = TransactionBuilder.CreateEmpty()
    .ValidUntil(NativeContract.Ledger.CurrentIndex(store) + 1)
    .AttachSystem(script => script.EmitDynamicCall(NativeContract.NEO.Hash, "transfer", UInt160.Zero, UInt160.Zero, 10, null))
    .AddAttributes(attr => attr.AddHighPriority())
    .AddSigner((signer, tx) =>
    {
        signer.Account(UInt160.Zero);
        signer.AddWitnessScope(WitnessScope.WitnessRules);
        signer.AddWitnessRule(WitnessRuleAction.Deny, rule =>
        {
            rule.AddCondition(condition =>
            {
                condition.And(and =>
                {
                    and.ScriptHash(UInt160.Zero);
                });
            });
        });
    })
    .AddWitness((witness, tx) =>
    {
        witness.AddInvocation(s => s.Emit(VM.OpCode.PUSHDATA1, [64, tx.Sign(kp, ProtocolSettings.Default.Network)]));
        witness.AddVerification(Contract.CreateSignatureRedeemScript(myPublicKey));
    })
    .Build();
```
## Applies To
- #3478

## Type of change

- [ ] Optimization (the change is only an optimization)
- [ ] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Wrote unit tests

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by vncoelho — COMMENTED** [review]
Looks good idea, @cschuchardt88 .

Maybe it would be good to create an issue to present the Neo.Builders with more details and where are the planned use cases.

**Comment by cschuchardt88** [comment]
@shargon @roman-khimov  @vncoelho @Jim8y 
Can you give me an example of a transaction with complex witness rules?

**Comment by cschuchardt88** [comment]
If anyone wants to test it, and give me some feedback on anything that needs to be improved.

**Comment by Jim8y** [comment]
> @shargon @roman-khimov @vncoelho @Jim8y Can you give me an example of a transaction with complex witness rules? For example how to interconnect in/with `Not`, `Or`, `And`, `ScriptHash`, `Group`, `Boolean` and `CalledBy*` conditions.

@superboyiii @Hecate2 do you have any idea?

**Comment by Hecate2** [comment]
> > @shargon @roman-khimov @vncoelho @Jim8y Can you give me an example of a transaction with complex witness rules? For example how to interconnect in/with `Not`, `Or`, `And`, `ScriptHash`, `Group`, `Boolean` and `CalledBy*` conditions.
> 
> @superboyiii @Hecate2 do you have any idea? 

On it. neo-fairy-client

**Comment by cschuchardt88** [comment]
You want to test it @Hecate2? Then give some feedback on the builder.

File: src/Neo/Builders/WitnessConditionBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+            SetConditionWithOtherConditions(condition);
+
+            return this;
+        }
+
+        public WitnessCondition Build()
+        {
+            if (_condition is null)
+                return new BooleanCondition() { Expression = false };
```
Comment (shargon) [inline src/Neo/Builders/WitnessConditionBuilder.cs:121]: Empty maybe should be true?

File: src/Neo/Builders/TransactionBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+using Neo.Network.P2P.Payloads;
+using Neo.VM;
+using System;
+
+namespace Neo.Builders
+{
+    public sealed class TransactionBuilder
+    {
+
```
Comment (shargon) [inline src/Neo/Builders/TransactionBuilder.cs:20]:
```suggestion
```

**Comment by Hecate2** [comment]
> You want to test it @Hecate2? Then give some feedback on the builder.

I think we can try to fill most properties of a tx automatically, unless the user specifies something else, or we do not have access to current chain enviromment. If we leave everything to be filled by the user, the complexity of building a tx is not reduced at all.
For example, the builder can try to get current block index to fill ValidUntil, and try to get current opened wallet to fill signer (may be difficult because wallets exists in Neo.Cli and RpcServer instead of Neo).
Ideally the user may just specify the contract call target and the params to build a new tx.

**Comment by cschuchardt88** [comment]
> > You want to test it @Hecate2? Then give some feedback on the builder.
> 
> I think we can try to fill most properties of a tx automatically, unless the user specifies something else, or we do not have access to current chain enviromment. If we leave everything to be filled by the user, the complexity of building a tx is not reduced at all. For example, the builder can try to get current block index to fill ValidUntil, and try to get current opened wallet to fill signer (may be difficult because wallets exists in Neo.Cli and RpcServer instead of Neo). Ideally the user may just specify the contract call target and the params to build a new tx.

I could create a new `Create` method in extensions or in class that would get additional environment data for building. How this is the 1st part of the builders. Next Im going to build `ApplicationEngine` builder, than integrate it in transaction builder for getting SystemFee. Same with NetworkFee and others.

**Review by Jim8y — APPROVED** [review]

File: src/Neo/Builders/WitnessBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            config(sb);
+            _verificationScript = sb.ToArray();
+            return this;
+        }
+
+        public WitnessBuilder AddVerification(byte[] bytes)
+        {
+            _verificationScript = bytes;
+            return this;
+        }
```
Comment (shargon) [inline src/Neo/Builders/WitnessBuilder.cs:56]: If it's already defined, it's remplaced, this doesn't happend with `AddWitness` for example

Comment (cschuchardt88) [reply src/Neo/Builders/WitnessBuilder.cs:56]: because you can have more than one witness.

Comment (cschuchardt88) [reply src/Neo/Builders/WitnessBuilder.cs]: `WitnessBuilder` is the builder for a single witness. `AddWitness` is adding multiple witnesses.

Comment (shargon) [reply src/Neo/Builders/WitnessBuilder.cs]: Throw an error?

Comment (cschuchardt88) [reply src/Neo/Builders/WitnessBuilder.cs]: builders normally don't do that. But I can.

Comment (cschuchardt88) [reply src/Neo/Builders/WitnessBuilder.cs]: Done!

**Review by Jim8y — APPROVED** [review]

**Pull request merged**